comment/src/comment.ts

```diff
@@ -0,0 +1,280 @@
+        })
+
+        return tr
+      }
+    }
+
+    return null
+  }
+
+  // Determines whether all selection ranges in `state` are block-commented.
+  isSelectionCommented(state: EditorState): { open: { pos: number, margin: number }, close: { pos: number, margin: number } }[] | null {
+    let result = []
+    for (const range of state.selection.ranges) {
+      const x = this.isRangeCommented(state, range)
+      if (x === null) return null
+      result.push(x)
+    }
+    return result
+  }
+
+  // Determines if the `range` is block-commented in the given `state`.
+  // The `range` must be a valid range in `state`.
+  isRangeCommented(state: EditorState, range: SelectionRange): { open: { pos: number, margin: number }, close: { pos: number, margin: number } } | null {
+    type SearchWithType = (this: string, searchString: string, pos?: number) => boolean
+    const search = (pos: number, searchString: string, searchWith: SearchWithType, d: 1 | -1, i: 1 | 0): { pos: number, margin: number } | null => {
+      const line = state.doc.lineAt(pos)
+      const str = line.content as string
+      const ss = eatSpace(str, pos - line.start - i, d)
+      return searchWith.call(str, searchString, pos + d * ss - line.start)
+        ? { pos: pos + d * ss, margin: ss == 0 ? 0 : 1 }
+        : null
+    }
+    const startsWithOpen = search(range.from, this.open, String.prototype.endsWith, -1, 1)
+    const endsWithClose = search(range.to, this.close, String.prototype.startsWith, 1, 0)
+    return startsWithOpen !== null && endsWithClose !== null
+      ? { open: startsWithOpen, close: endsWithClose }
+      : null
+  }
+
+  // Inserts a block comment in the given transaction `tr`.
+  insert(tr: Transaction, margin: string = " "): Transaction {
+    tr.forEachRange((range: SelectionRange, tr: Transaction) => {
+      const copen = new Change(range.from, range.from, tr.startState.splitLines(this.open + margin))
+      const cclose = new Change(range.to, range.to, tr.startState.splitLines(margin + this.close))
+      tr.change([copen, cclose])
+      const shift = (this.open + margin).length
+      return new SelectionRange(range.anchor + shift, range.head + shift)
+    })
+
+    return tr
+  }
+}
+
+// This class performs toggle, comment and uncomment
+// of line comments in languages that support them.
+// The `lineCommentToken` argument refer to the token of
+// which this `LineCommenter` is made up.
+export class LineCommenter {
+  lineCommentToken: string
+  margin: string
+  constructor(lineCommentToken: string, margin: string = " ") {
+    this.lineCommentToken = lineCommentToken
+    this.margin = margin
+  }
+
+  toggle(option: CommentOption, state: EditorState): Transaction | null {
+    const k = (range: SelectionRange): string => range.anchor + "," + range.head
+    const linesAcrossSelection: Line[] = []
+    const linesAcrossRange: { [id: string]: Line[]; } = {};
```

Comment:
Since you're using the same array of selection ranges throughout, why not index by range index? I.e. map `getLinesInRange` over `selection.ranges`, and on the second iteration index it again by the range's index.